edb/server/pgsql/metaschema.py

```diff
@@ -919,6 +978,109 @@ def __init__(self):
             text=self.__class__.text)
 
 
+class ArrayIndexWithBoundsFunction(dbops.Function):
+    """Get an array element or raise an out-of-bounds exception."""
+    text = '''
+        SELECT edgedb._raise_exception_on_null(
+            val[index],
+            'array_subscript_error',
+            'array index ' || (index - 1)::text || ' is out of bounds',
+            det
+        )
+    '''
+
+    def __init__(self):
+        super().__init__(
+            name=('edgedb', '_array_index'),
+            args=[('val', ('anyarray',)), ('index', ('bigint',)),
+                  ('det', ('text',))],
+            returns=('anyelement',),
+            volatility='immutable',
+            text=self.text)
+
+
+class StringIndexWithBoundsFunction(dbops.Function):
+    """Get a string character or raise an out-of-bounds exception."""
+    text = '''
+        SELECT edgedb._raise_exception_on_empty(
+            substr(val, index::int, 1),
+            'invalid_parameter_value',
+            'string index ' || (index - 1)::text || ' is out of bounds',
+            det
+        )
+    '''
+
+    def __init__(self):
+        super().__init__(
+            name=('edgedb', '_string_index'),
+            args=[('val', ('text',)), ('index', ('bigint',)),
+                  ('det', ('text',))],
+            returns=('text',),
+            volatility='immutable',
+            text=self.text)
+
+
+class JSONIndexWithBoundsFunction(dbops.Function):
+    """Get a JSON element or raise an out-of-bounds exception."""
+    text = r'''
+        SELECT
+            CASE WHEN jsonb_typeof(val) = 'object' THEN (
+                CASE WHEN pg_typeof(index)::text = 'text' THEN
```

Comment:
I'd use `IF` here for clarity.

Reply:
`IF` only exists in the imperative PLPGSQL, which we're trying to avoid in favor of pure SQL.